homeassistant/components/mqtt/client.py

```diff
@@ -669,6 +667,11 @@ def _process_client_subscriptions() -> list[tuple[int, int]]:
         subscriptions = self._pending_subscriptions
         self._pending_subscriptions = {}
 
+        for topic in subscriptions:
```

Comment:
This should probably be done inside of async_subscribe() not here, or there could be some weird race conditions.

Basically, async_subscribe should check for pending unsubscribes and remove them, and async_unsubscribe should check for pending subscribes and remove them (if it's the last subscription, like it currently does).  We should never have something on the pending subscribe AND pending unsubscribe list at the same time, but that can happen with:

1. Perform an async subscribe
2. Perform an async unsuscribe
3. Perform an async subscribe
-- Now have an entry on both lists.

Things probably still function as they should (I can't think of a situation where we would miss a subscribe or perform an unsubscribe when we don't want to), but it feels wrong if there is a possibility they're on both lists.

Reply:
It just makes sure that any pending unsubscribe will be canceled when it is sure a new pending subscribe was registered. A tests was added to prove the behaviour. The risk that we will have unsubscribes and subscribes at the same time is small.

Reply:
True - but if you move this code to the async_subscribe, it will be non-existent (the risk of being on both lists).  That said, as I said, I can't think of a scenario where this actually causes a problem, it just 'feels' wrong.

Reply:
I think that risk can be ignored as it very unlikely this will happen on the same topic, also because of the check that are done before subscribing and unsubscribing.

Reply:
I now now what you mean

Reply:
Moved the code to `async_subscribe` as you suggested.